Tutorials tab v1 (#1126)
astronomer/docs: tutorials/astro-python-sdk.md

---
title: 'Develop an ETL pipeline with the Astro Python SDK'

> sidebar_label: 'Astro Python SDK'

Author: "write ETL" seems a little casual even for the sidebar, what about "Write an ETL pipeline with the Astro SDK"? I know it's not as action-oriented, but I could also go for "Your first ETL DAG" like how Confluent does it here: 
![Screen Shot 2022-09-12 at 3 12 17 PM](https://user-images.githubusercontent.com/74574233/189676219-6e4f11eb-c5d7-450d-b85e-7fe7f7cda58e.png)

id: astro-python-sdk
description: 'Run an ETL pipeline with the Astro SDK.'
---

This tutorial demonstrates how to use the Astro Python SDK through a simple ETL pipeline that you can run on your local machine. The pipeline you build will:

- **Extract** a file from S3 into a Snowflake relational table.
- **Transform** that table in Snowflake.
- **Load** that transformed table into a reporting table in Snowflake.

## Assumed knowledge

To get the most out of this tutorial, make sure you have a knowledge of:

- Basic Python and SQL.
- Airflow fundamentals, such as writing DAGs and defining tasks.

## Prerequisites

To complete this tutorial, you need:

- An AWS S3 account with a [storage bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/GetStartedWithS3.html). If you don't already have an account, Amazon offers 5GB of free storage in S3 for 12 months. This should be more than enough for this tutorial.
- A Snowflake Enterprise account. If you don't already have an account, Snowflake has a [free Snowflake trial](https://signup.snowflake.com/) for 30 days.
- Python 3.

> - A local Airflow environment. You can use either [Apache Airflow](https://airflow.apache.org/docs/apache-airflow/stable/start/local.html) or [the Astro CLI](https://docs.astronomer.io/astro/cli/get-started) to run your environment. If you're using the Astro CLI, we recommend [creating a new Astro project](https://docs.astronomer.io/astro/create-project) for this tutorial.

Author: It depends whether the SDK team still plans to use this tutorial as a primary entry point for the SDK across all user contexts. It certainly makes sense to do that in the context of our docs, but maybe not in the context of the overall goal to get the SDK into as many hands as possible

## Step 1: Set up your data stores

> 1. On your local machine create a file named `orders_data_header.csv` with the following data:

Author: Nope, this file goes to S3 so there's no need for local storage.

    ```text
    order_id,customer_id,purchase_date,amount
    ORDER1,CUST1,1/1/2021,100
    ORDER2,CUST2,2/2/2022,200
    ORDER3,CUST3,3/3/2023,300
    ```

2. [Upload `orders_data_header.csv`](https://docs.aws.amazon.com/AmazonS3/latest/userguide/upload-objects.html) to your S3 bucket.
3. In Snowflake, create a new Worksheet and run the following SQL commands:

    ```sql
    CREATE DATABASE ASTRO_SDK_DB;
    CREATE WAREHOUSE ASTRO_SDK_DW;
    CREATE SCHEMA ASTRO_SDK_SCHEMA;
    ```

>     Ensure that you have `ACCOUNTADMIN` permissions for your newly created database.

Author: Opted to link out to Snowflake docs

---
title: 'Write an ETL pipeline with the Astro Python SDK'
sidebar_label: 'Write an ETL pipeline with the Astro SDK'
id: astro-python-sdk
description: 'Write an ETL pipeline for Amazon S3 and Snowflake with the Astro SDK.'
---

This tutorial demonstrates how to write an Extract, Transform, Load (ETL) pipeline on your local machine with the Astro Python SDK. The Astro SDK is maintained by Astronomer and simplifies the pipeline authoring process with native Python functions for common data orchestration use cases.

> The pipeline you build in this tutorial :

Reviewer: ```suggestion
The pipeline you build in this tutorial will:
```

- **Extract** a file from S3 into a Snowflake relational table.
- **Transform** that table in Snowflake.
- **Load** that transformed table into a reporting table in Snowflake.

## Assumed knowledge

To get the most out of this tutorial, make sure you have a knowledge of:

- Basic Python and SQL.
- Basic knowledge of [Amazon S3](https://aws.amazon.com/s3/getting-started/) and [Snowflake](https://docs.snowflake.com/en/user-guide-getting-started.html).
- Airflow fundamentals, such as writing DAGs and defining tasks. See [Get started with Apache Airflow](get-started-with-airflow.md).

## Prerequisites

To complete this tutorial, you need:

- An AWS S3 account with a [storage bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/GetStartedWithS3.html). If you don't already have an account, Amazon offers 5GB of free storage in S3 for 12 months. This should be more than enough for this tutorial.
- A Snowflake Enterprise account. If you don't already have an account, Snowflake has a [free Snowflake trial](https://signup.snowflake.com/) for 30 days.
- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).

## Step 1: Set up your data stores

> To run the DAG, you first need to load some arbitrary data into S3 and create a destination for the data in Snowflake.

Reviewer: ```suggestion
For this example, you first need to load a file into S3 and create a destination for the data in Snowflake. In this case, our file has three example orders with distinct customers, purchase dates, and amounts.
```

1. On your local machine create a file named `orders_data_header.csv` with the following data:

    ```text
    order_id,customer_id,purchase_date,amount
    ORDER1,CUST1,1/1/2021,100
    ORDER2,CUST2,2/2/2022,200
    ORDER3,CUST3,3/3/2023,300
    ```

2. [Upload `orders_data_header.csv`](https://docs.aws.amazon.com/AmazonS3/latest/userguide/upload-objects.html) to your S3 bucket.
3. In Snowflake, create a new [worksheet](https://docs.snowflake.com/en/user-guide/ui-worksheet.html) and run the following SQL commands:

    ```sql
    CREATE DATABASE ASTRO_SDK_DB;
    CREATE WAREHOUSE ASTRO_SDK_DW;
    CREATE SCHEMA ASTRO_SDK_SCHEMA;
    ```

    These commands create all of the required data stores for the load step of your pipeline. Ensure that you have `ACCOUNTADMIN` permissions for your newly created database.

## Step 2: Set up your Airflow environment

> 

Reviewer: ```suggestion

Now that we have a staging table in Snowflake and some example data ready to load, we need to set up a local Airflow environment for our DAG to run in. This is easy with the Astro CLI.
```

> 1. Run the following commands to create a new Astro project

Reviewer: ```suggestion
1. Create a new Astro project:
```

    ```sh
    $ mkdir astro-sdk-tutorial && cd astro-sdk-tutorial

>     $ astro dev init
>     ```

Reviewer: ```suggestion
    $ astro dev init
    ```

This command creates a few files that are required to run Apache Airflow. For this example, ignore these files for now.
```

Author: We have an "assumed knowledge" that includes the first Getting Started tutorial, so I don't think we need to get into the project info here.

> 2. Add the following line to the `requirements.txt` file of your Astro project to install the Python SDK and required dependencies for the tutorial:

Reviewer: ```suggestion
2. Add the following line to the `requirements.txt` file of your Astro project to install the Python SDK:
```

    ```text

>     astro-sdk-python[amazon,snowflake]>=0.11
>     ```

Reviewer: ```suggestion
    astro-sdk-python[amazon,snowflake]>=0.11
    ```
   This installs the SDK package as well as the Snowflake and Amazon providers, which are required for any DAG that loads data to or from Amazon S3 and Snowflake.
```

3. Add the following environment variables to the `.env` file of your project:

    ```text
    AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True
    AIRFLOW__ASTRO_SDK__SQL_SCHEMA=ASTRO_SDK_SCHEMA
    ```

    If you're running Airflow on macOS, additionally set the following environment variable [(background)](https://github.com/apache/airflow/issues/12808):

    ```shell
    export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES
    ```

4. Run the following command to start your project in a local environment:

    ```sh
    astro dev start
    ```

> ## Step 3: Create Airflow connections to data stores

Reviewer: ```suggestion
## Step 3: Create Airflow connections to S3 and Snowflake
```

Author: In service of keeping this as platform neutral as possible, I don't think we need to specify the tools in the titles of the steps

1. Open the Airflow UI at `http://localhost:8080/`
2. Go to **Admin** > **Connections**.
3. Create a new connection to AWS S3 with the following values:

    - Connection ID: `aws_default`
    - Connection type: `S3`
    - Extra: `{"aws_access_key_id": "<your_access_key>", "aws_secret_access_key": "<you_secret_access_key>"}`

4. Create a new connection to Snowflake with the following values

    - Connection Id: `snowflake_default`
    - Connection Type: `Snowflake`
    - Host: `https://<account>.<region>.snowflakecomputing.com/`. This is the URL where you can log into your Snowflake account.
    - Schema: `ASTRO_SDK_SCHEMA`
    - Login:
    - Password:
    - Account:
    - Database: `ASTRO_SDK_DB`
    - Region: (something like `us-east-1` or `us-central1.gcp`)
    - Role: `ACCOUNTADMIN`
    - Warehouse: `ASTRO_SDK_DW`

## Step 4: Create and populate some tables in Snowflake

Create some auxiliary tables in Snowflake and populate them with a small amount of data for this ETL example.

> 1. In your Snowflake Worksheet, create and populate a `customers_table`. You'll join this table with an orders table that you create with the Astro Python SDK:

Reviewer: ```suggestion
1. In your Snowflake Worksheet, create and populate a `customers_table`. You'll join this table with an orders table that you create with the Astro Python SDK:
```

Should we specify that this should be done in the DB and schema that we created in step 1?

    customers_table = Table(
        name=SNOWFLAKE_CUSTOMERS,
        conn_id=SNOWFLAKE_CONN_ID,
    )

    # Filter the orders data and then join with the customer table,
    # saving the output into a temporary table referenced by the Table instance `joined_data`
    joined_data = join_orders_customers(filter_orders(orders_data), customers_table)

    # Merge the joined data into the reporting table based on the order_id.
    # If there's a conflict in the customer_id or customer_name, then use the ones from
    # the joined data
    reporting_table = aql.merge(
        target_table=Table(
            name=SNOWFLAKE_REPORTING,
            conn_id=SNOWFLAKE_CONN_ID,
        ),
        source_table=joined_data,
        target_conflict_columns=["order_id"],
        columns=["customer_id", "customer_name"],
        if_conflicts="update",
    )

    # Transform the reporting table into a dataframe
    purchase_dates = transform_dataframe(reporting_table)

    # Delete temporary and unnamed tables created by `load_file` and `transform`, in this example
    # both `orders_data` and `joined_data`
    aql.cleanup()
```

> This DAG extracts the data you loaded into S3 into Airflow and runs a few simple SQL statements to clean the data, load it into a reporting table on Snowflake, and transform it into a dataframe so that you can print various table details to Airflow logs using Python.

Reviewer: ```suggestion
This DAG extracts the data you loaded into S3 and runs a few simple SQL statements to clean the data, load it into a reporting table on Snowflake, and transform it into a dataframe so that you can print various table details to Airflow logs using Python.
```

Much of this DAG's functionality comes from the Python functions that use task decorators from the Python SDK. See [How it works](#how-it-works) for more information about these decorators and the benefits of using the SDK for this implementation.  

## Step 6: Run the code

1. In the Airflow UI, you should see a DAG called `astro_orders`. Make it active by clicking the slider next to its name:

    ![toggle](/img/guides/unpause-dag.png)

2. Click the play button next to the DAG's name to run the DAG:

    ![trigger](/img/guides/trigger-dag.png)

3. Click the DAG's name to see how it ran in the **Grid** view:

    ![gridview](/img/guides/select-dag-grid-view.png)

> ## How does it work?

Reviewer: ```suggestion
## How it works
```

I think change `#how-it-works` to this?
